### PR TITLE
[ticket/14968] Distinguish between update/upgrade & show upgrade info in ACP index

### PR DESCRIPTION
**Pull request by marc1706**
Checklist:

- [x] Correct branch: master for new features; 3.2.x, 3.1.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master / 3.2.x](https://area51.phpbb.com/docs/master/coding-guidelines.html), [3.1.x](https://area51.phpbb.com/docs/31x/coding-guidelines.html)
- [x] Commit follows commit message [format](https://wiki.phpbb.com/Git#Commit_Messages)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-14968

### CODE REVIEW

File: phpBB/includes/acp/acp_main.php [hunk 1]
```diff
@@ -431,13 +431,23 @@ function main($id, $mode)
...
 		if ($auth->acl_get('a_board'))
 		{
+			/** @var \phpbb\version_helper $version_helper */
 			$version_helper = $phpbb_container->get('version_helper');
 			try
 			{
 				$recheck = $request->variable('versioncheck_force', false);
-				$updates_available = $version_helper->get_suggested_updates($recheck);
+				$updates_available = $version_helper->get_update_on_branch($recheck);
+				$upgrades_available = $version_helper->get_suggested_updates();
```
Comment (Nicofuma) [inline phpBB/includes/acp/acp_main.php]: Adds $recheck option?

File: phpBB/phpbb/version_helper.php [hunk 2]
```diff
@@ -201,6 +201,45 @@ public function get_latest_on_current_branch($force_update = false, $force_cache
...
+	 */
+	public function get_update_on_branch($force_update = false, $force_cache = false)
+	{
+		$versions = $this->get_versions_matching_stability($force_update, $force_cache);
+
+		$self = $this;
+		$current_version = $this->current_version;
+
+		// Filter out any versions less than to the current version
+		$versions = array_filter($versions, function($data) use ($self, $current_version) {
```
Comment (Nicofuma) [inline phpBB/phpbb/version_helper.php]: $self is not needed with PHP 5.4+

File: phpBB/phpbb/version_helper.php [hunk 3]
```diff
@@ -201,6 +201,45 @@ public function get_latest_on_current_branch($force_update = false, $force_cache
...
+			return $self->compare($data['current'], $current_version, '>=');
+		});
+
+		// Get the lowest version from the previous list.
+		return array_reduce($versions, function($value, $data) use ($self, $current_version) {
+			if ($value === null && $self->compare($data['current'], $current_version, '>='))
+			{
+
```
Comment (Nicofuma) [inline phpBB/phpbb/version_helper.php]: you can remove this empty line

File: tests/version/version_test.php [hunk 4]
```diff
@@ -332,4 +332,204 @@ public function test_get_latest_on_current_branch($current_version, $versions, $
...
+					),
+					'1.1'	=> array(
+						'current'		=> '1.1.1',
+					),
+				),
+				array(
+					'current'		=> '1.0.1',
+					'security'		=> '1.0.1',
+				),
+			),
```
Comment (Nicofuma) [inline tests/version/version_test.php]: It's the same test than the previous one

File: phpBB/includes/acp/acp_main.php (hunk 1 above)
Comment (marc1706) [inline phpBB/includes/acp/acp_main.php]: No. That would result in force fetching the latest version info twice. The first call will force fetch it and the second call can use that force fetched data.

File: phpBB/phpbb/version_helper.php (hunk 2 above)
Comment (marc1706) [inline phpBB/phpbb/version_helper.php]: But this is also for 3.1 which relies on 5.3.3.

File: tests/version/version_test.php (hunk 4 above)
Comment (marc1706) [inline tests/version/version_test.php]: Fixed below. Github caching is still displaying the old status with it saying `security --> 1.0.1` but it's supplying 1.0.0 now.

File: phpBB/includes/acp/acp_main.php (hunk 1 above)
Comment (Nicofuma) [inline phpBB/includes/acp/acp_main.php]: right

File: phpBB/phpbb/version_helper.php (hunk 2 above)
Comment (Nicofuma) [inline phpBB/phpbb/version_helper.php]: :cry:

File: phpBB/phpbb/version_helper.php [hunk 5]
```diff
@@ -201,6 +201,49 @@ public function get_latest_on_current_branch($force_update = false, $force_cache
...
+		$self = $this;
+		$current_version = $this->current_version;
+
+		// Filter out any versions less than to the current version
+		$versions = array_filter($versions, function($data) use ($self, $current_version) {
+			return $self->compare($data['current'], $current_version, '>=');
+		});
+
+		// Get the lowest version from the previous list.
+		$update_info = array_reduce($versions, function($value, $data) use ($self, $current_version) {
```
Comment (VSEphpbb) [inline phpBB/phpbb/version_helper.php]: ahem, bracket on newline :)

**Pull request merged**